Prefixed ulid source generator

marcin-dardzinski opened the pull request:
I suggest starting the reading from the second commit onwards, the first one is vendoring Ulid package (with some auto refactoring)

marcin-dardzinski committed:
Vendor Ulid from https://github.com/Cysharp/Ulid

marcin-dardzinski committed:
Implement ISpanFormattable and ISpanParsable for Ulid

marcin-dardzinski committed:
Add ulid backed source generated ids

marcin-dardzinski committed:
Address compiler warnings in Ulids

marcin-dardzinski committed:
Add Ulid property to strongly generated ids

marcin-dardzinski committed:
Normalized generated ids to have upper cased ulid part

Saancreed commented on Directory.Build.targets:
```diff
@@ -110,6 +110,8 @@
 
     <PackageReference Update="OpenTelemetry" Version="1.3.1" />
+
+    <PackageReference Update="NUlid" Version="1.3.2" />
```
I understand this is an alternative implementation used only for testing purposes?

marcin-dardzinski replied:
Yes

Saancreed commented on src/Domain/LeanCode.DomainModels/LeanCode.DomainModels.csproj:
```diff
@@ -1,3 +1,5 @@
 <Project Sdk="Microsoft.NET.Sdk">
-
+  <PropertyGroup>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
:thinking:

Let's see if we can get rid of it.

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+[DebuggerDisplay("{ToString(),nq}")]
+[System.Text.Json.Serialization.JsonConverter(typeof(UlidJsonConverter))]
+[System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
+public struct Ulid : IEquatable<Ulid>, IComparable<Ulid>, ISpanFormattable, ISpanParsable<Ulid>
```
`readonly`? Possibly also make it a `record`? Possibly also implement `IUtf8SpanFormattable`?

marcin-dardzinski replied:
Some marshal code was complaining after making it `readonly`

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+public struct Ulid : IEquatable<Ulid>, IComparable<Ulid>, ISpanFormattable, ISpanParsable<Ulid>
+{
+    // https://en.wikipedia.org/wiki/Base32
+    private static readonly char[] Base32Text = "0123456789ABCDEFGHJKMNPQRSTVWXYZ".ToCharArray();
```
I'd make all the private readonly collections immutable if possible.

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+        31
+    };
+
+    private static readonly DateTimeOffset UnixEpoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
```
Consider using [`DateTimeOffset.UnixEpoch`](https://learn.microsoft.com/en-us/dotnet/api/system.datetimeoffset.unixepoch?view=net-8.0) instead?

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+
+    [IgnoreDataMember]
+    [SuppressMessage("?", "CA1819", Justification = "Fresh array is allocated per call")]
+    public byte[] Random =>
```
I'd change return type to `ROS<byte>` or `ROM<byte>` or something similar and reimplement it as a read-only view over current instance's data. The copy is probably not always needed.

marcin-dardzinski replied:
Giving it a second thought I don't think there's a safe way to convert it into ROS. If the method is not inlined the stack frame that we'll be pointing to will be invalidated. Maybe a static method with `ref` arg would help.

Still, I think it might be risky, say you get a span to the id in some array and then swap it with some other id.

On the other hand, nobody uses this property now so we can change it into `TryGetRandom(Span<byte>)` or just get rid of it :upside_down_face:

Dragemil commented on src/Domain/LeanCode.DomainModels.Generators/IdSource.cs:
```diff
@@ -140,6 +144,114 @@ public static bool IsValid([NotNullWhen(true)] string? v)
+        }
+
+        public bool Equals({{data.TypeName}} other) => Value.Equals(other.Value, StringComparison.Ordinal);
+        public int CompareTo({{data.TypeName}} other) => string.Compare(Value, other.Value, StringComparison.Ordinal);
```
I guess you could compare just the `Ulid` parts here, which might've been more efficient.

marcin-dardzinski replied:
Nah, string comparison is vectorized anyway, little to zero gain but a readability cost

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+            buffer[4] = timestamp1;
+            buffer[5] = timestamp0; // [6], [7] = 0
+
+            var timestampMilliseconds = Unsafe.As<byte, long>(ref MemoryMarshal.GetReference(buffer));
```
What about endianness here? Wouldn't it be safer to use something like `BinaryPrimitives.ReadInt64LittleEndian` instead?

Dragemil commented on src/Domain/LeanCode.DomainModels.Generators/IdSource.cs:
```diff
@@ -140,6 +144,114 @@ public static bool IsValid([NotNullWhen(true)] string? v)
+            return TryDeconstruct(v.AsSpan(), out _);
+        }
+
+        public bool Equals({{data.TypeName}} other) => Value.Equals(other.Value, StringComparison.Ordinal);
```
You here could compare only `Ulid` parts.

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+    internal Ulid(long timestampMilliseconds, XorShift64 random)
+        : this()
+    {
+        // Get memory in stack and copy to ulid(Little->Big reverse order).
+        ref var firstByte = ref Unsafe.As<long, byte>(ref timestampMilliseconds);
```
Again, what if `long` is already BE?

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+    public Ulid(Guid guid)
+    {
+        Span<byte> buf = stackalloc byte[16];
+        MemoryMarshal.Write(buf, ref guid);
+        if (BitConverter.IsLittleEndian)
```
I think `Guid.TryWriteBytes` handles some of that nonsense for you, even if that would make the code a bit slower on BE machines.

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+        try
+        {
+            TryWriteBytes(buffer);
+            return Convert.ToBase64String(buffer, options);
```
`Convert.ToBase64String` has an overload that accepts `ROS<byte>` so you could avoid this by stackalloc'ing 16 bytes, writing there and converting that.

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+    public override unsafe int GetHashCode()
+    {
+        // Simply XOR, same algorithm of Guid.GetHashCode
+        fixed (void* p = &this.timestamp0)
```
Uhh okay, this is a bit mindbending. I think we could get away with `Unsafe.ReadUnaligned(Unsafe.Add(…))` but I'll have to verify that.

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+    public unsafe bool Equals(Ulid other)
+    {
+        // readonly struct can not use Unsafe.As...
+        fixed (byte* a = &this.timestamp0)
```
How about we just created a `ROS<byte>` over both instances and then compared their content?

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -0,0 +1,884 @@
+
+    public int CompareTo(Ulid other)
+    {
+        if (this.timestamp0 != other.timestamp0)
```
And here, reinterpret memory as `ROS<byte>` and run compare on that with `SequenceCompareTo`. Might be a bit slower, but I'd be okay with that.

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/UlidRandomProvider.cs:
```diff
@@ -0,0 +1,71 @@
+    private static XorShift64 CreateXorShift64()
+    {
+        using var rng = RandomNumberGenerator.Create();
+        Span<byte> buffer = stackalloc byte[sizeof(ulong)];
```
Isn't `sizeof` unsafe? :open_mouth:

If it is, you could replace this with `Unsafe.SizeOf<ulong>()` or just dump literal `8` there.

Saancreed reviewed (commented):
I have a few suggestions :see_no_evil:

By the way, didn't we want to keep vendored `Ulid` internal or was that just my imagination?

marcin-dardzinski reviewed (commented):
>By the way, didn't we want to keep vendored Ulid internal or was that just my imagination?

It has to be `public` to work with source generators.

Noted all the _modernization_ comments - I'll create a sub PR addressing them, they'll be better trackable.

marcin-dardzinski committed:
Make private ulid arrays immutable, use library provided unix epoch start

marcin-dardzinski committed:
Implement `IUtf8SpanFormattable`, remove `TryWriteStringify` methods

marcin-dardzinski committed:
Rewrite `Equals`/`CompareTo`/`GetHashCode` into span based methods

jakubfijalkowski reviewed (approved):
:+1: to all the comments @Saancreed left. Overall LGTM.

marcin-dardzinski committed:
Don't assume little endian in dates when constructing ulid

marcin-dardzinski committed:
Simplify guid interop, add more tests

marcin-dardzinski committed:
Ulid code review fixes

1. Aggressively inline Ulid.AsSpan()
2. Rename LengthInTextElements for Ulid string length
3. Explain endianness swapping in Guid conversions
